### DISCUSSION TITLE
Feature request: Scoped Labels

### DISCUSSION THREAD

**Discussion by itshypax** [community/community#16682] [Projects and Issues]
This pretty much regards labels of pull requests and issues.

GitLab uses a feature where users can use "scoped labels" which splits the labels apart. Where users create a label and just put in `::` in between the parts that they want to split.

So, for example, it could be used to create different categories of issues with `issue::translation`, priorities `priority::high` or versioning `release::V1.0.0`. I believe it could have many more use cases and could generally help improve on managing/sorting pull requests or issues. 

Visual example (from Gitlab):
![Scoped Labels](https://docs.gitlab.com/ee/user/project/img/labels_key_value_v13_5.png)

**Comment by antoine-pous** [comment]
I came to ask this feature too, but scoped labels are not just a visual enhancement. They work as group labels, so if you have `Status::In progress` and then add `Status::Done` GitLab will remove `Status::In Progress` considering we can only have one label from `Status` scope.

It miss a bit of configuration on GitLab, like:
- Being able to setup label behavior (eg: `Status::*` is unique, but we can have many `Service::*` labels)
- Setting up a color directly to the scope instead of doing it for each label

Currently this feature on GitLab is incredibkly valuable and helpful, this is one of the reasons my company is on it.

**Reply by tayloramurphy** [reply]
Coming from GitLab over to GitHub this is definitely something I miss. My hunch is that GitHub would probably prefer to support this via a workflow or custom github action rather than as a native feature.

**Reply by LukeSavefrogs** [reply]
This is something Github lacks of...

**Comment by alystair** [comment]
This is the primary reason I still actively use GitLab for managing many orgs/projects. Custom action/workflow might work but they should provide an explanation of how to implement it

**Comment by pd93** [comment]
I would really love to see this added to GitHub, even if its a visual-only thing. It's one of the few features from GitLab I still miss when organising a project. I get that we can emulate the stateful labels using workflows, but there is no built-in protection against manually adding two labels with the same scope and things can get out of sync.

**Comment by lukens** [comment]
Just nudging this discussion again, as there's barely a day goes by without me wishing for this feature.

**Comment by dicamilloAR** [comment]
how do you manage this now? I am going from GitLab to GitHub and I am curious to see if anyone has dealt with it in any way that was helpful/manageable.

**Reply by itshypax** [reply]
Unfortunately, I don't think there's any way to emulate this feature at the moment.

**Reply by pd93** [reply]
There's no reason why you can't create a workflow that does this for you. In [go-task/task](https://github.com/go-task/task), we have [some automation](https://github.com/go-task/task/blob/main/.github/workflows/issue-experiment.yml) that uses [github-script](https://github.com/actions/github-script) for our experiments workflow that is based off "scoped" labels. We name all our labels in the format: `<scope>: <state>` and then run the workflow when a label is added to something.

Currently, we're just using it to add comments to issues, but there is no reason why it couldn't be extended to also remove any other labels with the same scope. It's not perfect, but I think it's as close as you can get right now.

**Comment by leofang** [comment]
Bumping this thread in 2024. This is perhaps the only missing feature compared to gitlab...

**Comment by mahtaran** [comment]
With the work being done now in #139935, #139932, #139933, now seems like a more logical time than ever to have another look at this.

**Reply by itshypax** [reply]
Unfortunately, it doesn't look like GitHub will ever implement this feature.

**Reply by mahtaran** [reply]
So it seems, but it would be awesome to get any kind of official response (even if it's a confirmation that it's nowhere on the roadmap) for that.

**Comment by LeaVerou** [comment]
OMG yes, I've been wishing for this for years. I have SO MANY `Key: value` labels that I transfer from repo to repo. But labels seem like a workaround for what is fundamentally fields. So perhaps that could be a solution, with a way to migrate from labels? I filed https://github.com/orgs/community/discussions/149754 earlier today for this exact thing.

**Reply by itshypax** [reply]
This request or suggestions has been open for so long and never been answered or even in any way responded to by GitHub that I really believe that there is no real chance that it will ever happen. Sadly. 🥲

**Comment by Ernest314** [comment]
In case it matters, Gitea has also had this feature for quite awhile now (https://docs.gitea.com/usage/labels#scoped-labels).

**Reply by itshypax** [reply]
I would love to see this feature in GitHub. It really is one of the biggest things missing for versioning etc. for me. I really would love to have this feature here. I loved it when I used in GitLab for a project back then.

**Reply by Ernest314** [reply]
It was so useful when I used it in GitLab that I just *assumed* GitHub must've had the feature already (and I'd missed it somehow)

**Comment by sackville** [comment]
I'll add this: GitHub covers many of the use cases for scoped labels with fields on issues and product items. The issue `Type` field acts just like a scoped label, as do the project `Status`, `Priority`, and `Size` fields. You can configure their options, and they are each mutually exclusive.

HOWEVER, that's just excellent functionality with a limited reach. Those fields only cover a handful of use cases (clearly the most popular and important ones). They solve the core problem for those use cases: Categorization of issues that is configurable, easy to use, visually clear, and quick to interact with.

_Scoped labels_ extend the solution by allowing teams to choose the dimensions along which their issues will be categorized, and they do so without impacting the dimensions used by other organizations or teams. This allows a team to find and focus on the differentiations between issues that matter most to their org, team or development process. If they look for issues according to functional expertise, they may create a `function::` scope, with labels for `function::design`, `function::development`, or `function::test`. If they manage releases with labels they can find every issue labeled with a release by creating a `release::` scope and search on `release::*`. If the team organizes content by theme, a `theme::` scope and labels like `theme::devops`, `theme::security`, or `theme::cat videos` would make organizing and finding that content a cinch. All without any custom fields!

I'm pretty sure I'm not the only one looking into a way to implement scoped labels within GitHub without direct support from GitHub. I imagine the number of interested users would be easy to find by just looking up how many labels have a `<foo>::<bar>` or `<foo>: <bar>` pattern. All of us would really appreciate native support for scoped labels.

**Comment by beingminimal** [comment]
@mariorod @ashtom Please consider this request. Thanks

**Comment by labudis** [comment]
We shipped [Issue Fields](https://github.com/orgs/community/discussions/189141), which covers many of the use cases described here. You can now define org-wide typed fields (single-select, text, number, date) that apply across all repositories. For example, Priority, Effort, or any custom category your team needs. Fields are searchable, filterable, and available in Projects.

This isn't scoped labels per se, but it addresses the core need: structured, consistent metadata on issues without label workarounds.

**Reply by lukens** [reply]
Are these only org wide? I don't want to have to get org wide agreement on such things!

**Reply by doyled-it** [reply]
Not quite the same. Issue labels carry over from issues into MRs in GitLab. PRs don't have `PR Fields`.

**Reply by soloturn** [reply]
@labudis label is also a field, can be queried, and, best of all, we have it at pull requests as well. which makes displaying pull requests in a board easier.

**Reply by jon-nfc** [reply]
> We shipped [Issue Fields](https://github.com/orgs/community/discussions/189141), which covers many of the use cases described here. You can now define org-wide typed fields (single-select, text, number, date) that apply across all repositories. For example, Priority, Effort, or any custom category your team needs. Fields are searchable, filterable, and available in Projects.
> 
> This isn't scoped labels per se, but it addresses the core need: structured, consistent metadata on issues without label workarounds.

Why has this been marked as the answer?

OP **clearly** asked for **scoped labels**. your answer to the contrary are not scoped labels.

@itshypax can you unmark this as the answer so it can be seen as not completed. or @labudis can you unmark this as the answer being the one who **should not have** marked this as the answer.

Like everyone else in this discussion **"We want scoped labels!!!!! not fields"**

**Reply by itshypax** [reply]
@jon-nfc I never questioned what labudis said, but I've unmarked it as answered.

**Comment by LeaVerou** [comment]
Is there any estimate of when these will ship to the public? I'm working on a lot of unreleased projects and don't feel comfortable asking for them to be enabled on these orgs here but would love to be able to use the functionality.
